Update blackjack v2 guide (#16341)
MystenLabs/sui: docs/content/guides/developer/app-examples/blackjack.mdx

:::info

For more details on building a back-end and deploying to Sui, check out the [Coin Flip app example](./coin-flip.mdx).

You can also find the [full repository for this example here](https://github.com/MystenLabs/blackjack-sui).

:::

## Gameplay

In this single-player version of blackjack, the player competes against a dealer, which is automated by the system. The dealer is equipped with a public BLS key that plays a central role in the game's mechanics. The dealer's actions are triggered by HTTP requests to serverless functions. Players generate randomness for the game by interacting with their mouse on the screen, after which they place their bet to start the game. Upon initiating the game, a request is made to the dealer's back-end, which processes it by signing and subsequently dealing two cards to the player and one to themselves.

The player has the options to 'Hit' or 'Stand.' Selecting 'Stand' triggers the dealer to draw cards until the total reaches 17 or higher. After the dealer stops, the smart contract steps in to compare the totals and declare the winner. On the other hand, choosing 'Hit' prompts the dealer to draw an additional card for the player.

> Note that more complex Blackjack rules, such as splitting, are considered out of scope for this example, and are therefore not implemented.

Reviewer: ```suggestion
Note that more complex blackjack rules, such as splitting, are considered out of scope for this example, and are therefore not implemented.
```

| POST        | /api/games/{id}/deal  | Executes the initial deal transaction after the game creation | txDigest                                          |
| POST        | /api/games/{id}/hit   | Executes a hit move                                           | txDigest, id of corresponding HitRequest object   |
| POST        | /api/games/{id}/stand | Executes a stand move                                         | txDigest, id of corresponding StandRequest object |

### Need of usage of waitForTransactionBlock

An interesting aspect of developing a Dapp on Sui, that is coupled to using a full node with/without a load balancer, and requires attention by the developer, is the occurrence of read-after-write and write-after-write cases.

#### Initial Deal Transaction

As an example, in our Blackjack game, just after the create-game transaction that is executed by the user, the dealer executes the initial-deal transaction. This one, accepts as argument and modifies the game object, meaning that we are using on an object that was just created.

> To ensure that the game object is available in the full node that the dealer is using, we need to call waitForTransactionBlock after the create-game transaction.

Reviewer: ```suggestion
To ensure that the game object is available in the Full node that the dealer is using, we need to call waitForTransactionBlock after the create-game transaction.
```

> #### Hit and Stand Transactions

Reviewer: ```suggestion
#### Hit and stand transactions
```

> In the same way, every time we re-fetch the game object in the front-end, we need to make sure that the previous transaction that modified the game object is already available in the full node.

Reviewer: ```suggestion
In the same way, every time you re-fetch the game object in the frontend, make sure that the previous transaction that modified the game object is already available in the Full node.
```

> This leads to the need of exchanging the txDigest between the front-end and the back-end, and use waitForTransactionBlock on each write-after-write or read-after-write case.

Reviewer: ```suggestion
This leads to the need of exchanging the `txDigest` between the frontend and the backend, and use `waitForTransactionBlock` on each write-after-write or read-after-write case.
```

> ## Front-end

Reviewer: ```suggestion
## Frontend
```

> The [`page` component](https://github.com/MystenLabs/blackjack-sui/blob/main/app/src/app/page.tsx), a central element of the blackjack game's front-end module, is structured to create an interactive and responsive gaming experience. Written in React, it integrates several features and functions to handle the game's logic and user interactions effectively.

Reviewer: ```suggestion
The [`page` component](https://github.com/MystenLabs/blackjack-sui/blob/main/app/src/app/page.tsx), a central element of the blackjack game's frontend module, is structured to create an interactive and responsive gaming experience. Written in React, it integrates several features and functions to handle the game's logic and user interactions effectively.
```

> ### Directories Structure

Reviewer: ```suggestion
### Directories structure
```

> The front-end is a NextJS project, that follows the NextJS App Router [project structure](https://nextjs.org/docs/app/building-your-application/routing).

Reviewer: ```suggestion
The frontend is a NextJS project, that follows the NextJS App Router [project structure](https://nextjs.org/docs/app/building-your-application/routing).
```

> The main code of the front-end lies under the [app/src/](https://github.com/MystenLabs/blackjack-sui/blob/main/app/src/app/) directory.

Reviewer: ```suggestion
The main code of the frontend lies under the [app/src/](https://github.com/MystenLabs/blackjack-sui/blob/main/app/src/app/) directory.
```

The main sub-directories are:
  - [app/](https://github.com/MystenLabs/blackjack-sui/blob/main/app/src/app/): The main code of the pages and the API endpoints.
  - [components/](https://github.com/MystenLabs/blackjack-sui/blob/main/app/src/components): The reusable components of the app, organized in sub-directories.
  - [hooks/](https://github.com/MystenLabs/blackjack-sui/blob/main/app/src/hooks): The custom hooks used in the app.
  - [helpers/](https://github.com/MystenLabs/blackjack-sui/blob/main/app/src/helpers), [utils/](https://github.com/MystenLabs/blackjack-sui/blob/main/app/src/utils), [lib/](https://github.com/MystenLabs/blackjack-sui/blob/main/app/src/lib): Multiple helper functions and utilities.
  - [types/](https://github.com/MystenLabs/blackjack-sui/blob/main/app/src/types): The types/interfaces used in the app.
  - [styles/](https://github.com/MystenLabs/blackjack-sui/blob/main/app/src/styles): The global css files to style our app.

### Components and custom hooks for state management

> - **Custom Hooks:** In order to keep the code as structured as possible, multiple custom hooks are utilized to manage the complex state of the gameboard at each step. The [useBlackjackGame](https://github.com/MystenLabs/blackjack-sui/blob/main/app/src/hooks/useBlackjackGame.ts) custom hook encapsulates the game state and logic, exposing all the required information (with fields such as `game`, `isInitialDealLoading`), and the required functionality (with methods such as `handleCreateGame` and `handleHit`) to display and play the game. Multiple additional custom hooks, such as [useCreateBlackjackGame](https://github.com/MystenLabs/blackjack-sui/blob/main/app/src/hooks/useCreateBlackjackGame.ts), and [useMakeMoveInBlackjackGame](https://github.com/MystenLabs/blackjack-sui/blob/main/app/src/hooks/useMakeMoveInBlackjackGame.ts) are encapsulating their own piece of state and logic to make the code readable and maintainable.

Reviewer: ```suggestion
- **Custom Hooks:** To keep the code as structured as possible, multiple custom hooks are utilized to manage the complex state of the game board at each step. The [useBlackjackGame](https://github.com/MystenLabs/blackjack-sui/blob/main/app/src/hooks/useBlackjackGame.ts) custom hook encapsulates the game state and logic, exposing all the required information (with fields such as `game`, `isInitialDealLoading`), and the required functionality (with methods such as `handleCreateGame` and `handleHit`) to display and play the game. Multiple additional custom hooks, such as [useCreateBlackjackGame](https://github.com/MystenLabs/blackjack-sui/blob/main/app/src/hooks/useCreateBlackjackGame.ts), and [useMakeMoveInBlackjackGame](https://github.com/MystenLabs/blackjack-sui/blob/main/app/src/hooks/useMakeMoveInBlackjackGame.ts) are encapsulating their own piece of state and logic to make the code readable and maintainable.
```

### Differences

> - **Game Mechanics and Complexity:** Blackjack is a more complex game with multiple actions (hit, stand, deal) and state updates, requiring a more dynamic front-end. In contrast, Satoshi Coin Flip has a simpler mechanic centered around a single bet and guess outcome.

Reviewer: ```suggestion
- **Game Mechanics and Complexity:** Blackjack is a more complex game with multiple actions (hit, stand, deal) and state updates, requiring a more dynamic frontend. In contrast, Satoshi Coin Flip has a simpler mechanic centered around a single bet and guess outcome.
```

- **User Interface (UI) Complexity:** The Blackjack game involves a more intricate UI to display cards, manage game states, and player interactions. Satoshi Coin Flip, being simpler in gameplay, requires a less complex UI.

> - **Back-end Processing:** In Blackjack, the dealer is automated (the machine), and the player's actions directly influence game outcomes. In the Coin Flip game, the house (smart contract) plays a more passive role, primarily in initializing and finalizing the game based on the player's guess.

Reviewer: ```suggestion
- **Backend Processing:** In Blackjack, the dealer is automated (the machine), and the player's actions directly influence game outcomes. In the Coin Flip game, the house (smart contract) plays a more passive role, primarily in initializing and finalizing the game based on the player's guess.
```

> - **Module Structure and Focus:** The Blackjack game focuses more on front-end interactions and real-time updates. The Satoshi Coin Flip game, delves into back-end logic, with structures like `HouseCap` and `house_data` for initializing and managing game data securely on the blockchain.

Reviewer: ```suggestion
- **Module Structure and Focus:** The Blackjack game focuses more on frontend interactions and real-time updates. The Satoshi Coin Flip game, delves into backend logic, with structures like `HouseCap` and `house_data` for initializing and managing game data securely on the blockchain.
```
